llama_deploy/message_queues/solace/solace.py

```diff
@@ -0,0 +1,271 @@
+"""Solace Message Queue."""
+
+import asyncio
+import json
+import time
+import threading
+from string import Template
+from logging import getLogger
+from typing import Any, Dict, List, Literal, TYPE_CHECKING
+
+from pydantic import BaseModel, Field
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+from llama_deploy.message_queues.base import BaseMessageQueue
+from llama_deploy.messages.base import QueueMessage
+from llama_deploy.message_consumers.base import BaseMessageQueueConsumer, StartConsumingCallable
+
+# if TYPE_CHECKING:
+from solace.messaging.errors.pubsubplus_client_error import PubSubPlusClientError, IllegalStateError
+from solace.messaging.resources.topic_subscription import TopicSubscription
+from solace.messaging.resources.topic import Topic
+from solace.messaging.publisher.persistent_message_publisher import PersistentMessagePublisher, \
+    MessagePublishReceiptListener
+from solace.messaging.receiver.message_receiver import MessageHandler, InboundMessage
+from solace.messaging.receiver.persistent_message_receiver import PersistentMessageReceiver
+from solace.messaging.config.retry_strategy import RetryStrategy
+from solace.messaging.messaging_service import MessagingService
+from solace.messaging.resources.queue import Queue
+from solace.messaging.config.missing_resources_creation_configuration import MissingResourcesCreationStrategy
+
+from .boot import BootSolace
+
+# Constants
+MAX_SLEEP = 10
+QUEUE_TEMPLATE = Template('Q/$iteration')
+
+# Configure logger
+logger = getLogger(__name__)
+
+class MessagePublishReceiptListenerImpl(MessagePublishReceiptListener):
+    """Message publish receipt listener for Solace message queue."""
+    def __init__(self):
+        self._publish_count = 0
+        self._lock = threading.Lock()
+
+    @property
+    def get_publish_count(self):
+        return self._publish_count
+
+    def on_publish_receipt(self, publish_receipt: 'PublishReceipt'):
+        with self._lock:
+            self._publish_count += 1
+            logger.info(f"\tMessage: {publish_receipt.message}\n"
+                  f"\tIs persisted: {publish_receipt.is_persisted}\n"
+                  f"\tTimestamp: {publish_receipt.time_stamp}\n"
+                  f"\tException: {publish_receipt.exception}\n")
+            if publish_receipt.user_context:
+                logger.info(f'\tUsercontext received: {publish_receipt.user_context.get_custom_message}')
+
+class MessageHandlerImpl(MessageHandler):
+    """Message handler for Solace message queue."""
+    def __init__(self, consumer: BaseMessageQueueConsumer):
+        self.consumer = consumer
+
+    def on_message(self, message: InboundMessage):
+        topic = message.get_destination_name()
```

Comment:
Somewhere we need to acknowledge messages back to the broker. If you don't, they will remain on the broker's queue forever.